### PR TITLE
[stable/traefik] Add support for proxyProtocol.

### PR DESCRIPTION
**Pull request by timstoop**
This adds support for proxyProtocol, which allows traefik to present actual user IPs when behind a loadbalancer like EBS.

### CODE REVIEW

**Comment by unguiculus** [comment]
cc @krancour

File: stable/traefik/templates/configmap.yaml [hunk 1]
```diff
@@ -30,6 +39,15 @@ data:
...
       address = ":443"
+      {{- if .Values.proxyProtocol.enabled }}
+      [entryPoints.http.proxyProtocol]
+        trustedIPs = [
+          {{- range $idx, $element := .Values.proxyProtocol.trustedIPs }}
+              {{- if $idx }}, {{ end }}
+              {{- $element | quote }}
+          {{- end -}}
+          ]
+      {{- end }}
```
Comment (krancour) [inline stable/traefik/templates/configmap.yaml:51]: Can this bit go into a helper or something since it's duplicated in two places?

File: stable/traefik/Chart.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v1
 name: traefik
-version: 1.14.2
+version: 1.14.3
```
Comment (krancour) [inline stable/traefik/Chart.yaml]: Since this adds a feature, it should be the minor version number that is incremented.

**Comment by krancour** [comment]
Love the feature. Mad one comment about extracting some duplicated logic into a helper.

This will also need some updates to the README. There's a table in there that goes over all config options in detail. And It wouldn't actually hurt to have a who separate section of the documentation dedicated to this-- or at least point to the relevant Traefik docs. PROXY protocol is something that needs to be enabled on _both_ the proxy and the target to work. If applied in one place and not the other things will break. Just need to be sure people are aware.

And again-- I _love_ this addition.

File: stable/traefik/templates/configmap.yaml (hunk 1 above)
Comment (timstoop) [reply stable/traefik/templates/configmap.yaml:51]: I'd love to, but I'm afraid I'm going to need some guidance with that.

File: stable/traefik/README.md [hunk 3]
```diff
@@ -98,6 +98,8 @@ The following tables lists the configurable parameters of the Traefik chart and
 | `rbac.enabled`                  | Whether to enable RBAC with a specific cluster role and binding for Traefik | `false`                            |
 | `nodeSelector`                  | Node labels for pod assignment                                       | `{}`                                      |
 | `tolerations`                   | List of node taints to tolerate                                      | `[]`                                      |
+| `proxyProtocol.enabled`         | Enable receiving the Proxy protocol instead http or https            | `false`                                   |
+| `proxyProtocol.trustedIPs`      | List of trusted ip (ranges) from which to allow Proxy protocol traffic | `["10.0.0.0/8"]`                        |
```
Comment (krancour) [inline stable/traefik/README.md]: I would reword the descriptions above for better clarity...

Propose changing

> Enable receiving the Proxy protocol instead http or https

to just:

> Enable PROXY protocol support.

Maybe add a link to the section where you elaborate on it?

Propose changing

> List of trusted ip (ranges) from which to allow Proxy protocol traffic

to

> List of proxy IPs (CIDR ranges) trusted to accurately convey the end-user IP.

Comment (krancour) [reply stable/traefik/README.md]: The default of `10.0.0.0/8` isn't going to work for lots and lots of people. It might be better to _not_ have a default and make chart install fail if it's not defined. This can be done with the some conditional logic and the [fail function from Sprig](http://masterminds.github.io/sprig/flow_control.html) (which Helm supports).

File: stable/traefik/README.md [hunk 4]
```diff
@@ -179,3 +181,7 @@ dashboard:
     basic:
       test: $apr1$H6uskkkW$IgXLP6ewTrSuBkTrqE8wj/
 ```
+
+### Proxy Protocol
+
+Normally, when traefik lives behind a LoadBalancer (like an AWS ELB), incoming traffic is all originating from that LoadBalancer, from traefik's point of view.  By setting the ELB to `Proxy` ([via an annotation](https://kubernetes.io/docs/concepts/services-networking/service/#proxy-protocol-support-on-aws)) and telling traefik to expect the Proxy protocol, the ELB acts as a transparant proxy, ie. traefik will pass incoming traffic along like it originated from the actual visitor instead of the ELB. Basically, you get to see the actual remote IP address of the client that's making the request.
```
Comment (krancour) [inline stable/traefik/README.md]: s/traefik/Traefik/

Comment (krancour) [reply stable/traefik/README.md]: This section might need more rewording for clarity.

File: stable/traefik/templates/configmap.yaml (hunk 1 above)
Comment (krancour) [reply stable/traefik/templates/configmap.yaml:51]: Have a look at `_helpers.tpl` for examples of some helpers that are already used in this chart.

Comment (timstoop) [reply stable/traefik/templates/configmap.yaml:51]: I did, there's only one in there and it does a string replace. I'm looking for an example of how to replace an entire block of code. I'll keep looking in the other charts (tomorrow, it's 1.30am here).

**Comment by unguiculus** [comment]
ping @timstoop

**Comment by timstoop** [comment]
Apologies, I've been very busy :( I'll pick this up again next week, after Christmas.

**Comment by timstoop** [comment]
Darn it, this is not at all what I intended to do... Just wanted to rebase my branch! Closing this one and creating a new pull request, with all the comments resolved as much as I can.
